connector/connect/server/src/main/scala/org/apache/spark/sql/connect/planner/SparkMLPlanner.scala

```diff
@@ -0,0 +1,402 @@
+        lang.Long.valueOf(protoArg.getInt64Value)
+      case ArgValue.ArgTypeCase.FLOAT_VALUE =>
+        lang.Float.valueOf(protoArg.getFloatValue)
+      case ArgValue.ArgTypeCase.DOUBLE_VALUE =>
+        lang.Double.valueOf(protoArg.getDoubleValue)
+      case ArgValue.ArgTypeCase.STRING_VALUE =>
+        protoArg.getStringValue
+      case ArgValue.ArgTypeCase.BOOL_VALUE =>
+        lang.Boolean.valueOf(protoArg.getBoolValue)
+      case ArgValue.ArgTypeCase.RELATION =>
+        loadInputRelation(sessionHolder, protoArg.getRelation)
+      case ArgValue.ArgTypeCase.REMOTE_OBJECT =>
+        sessionHolder.serverSideObjectManager.getObject(protoArg.getRemoteObject.getId)
+      case ArgValue.ArgTypeCase.LIST =>
+        val protoValues = protoArg.getList.getElementList.asScala.toList
+        protoValues.map { x => parseArg(x, sessionHolder) }
+      case ArgValue.ArgTypeCase.MAP =>
+        val protoMap = protoArg.getMap.getMapMap().asScala.toMap
+        protoMap.mapValues(x => parseArg(x, sessionHolder))
+      case _ =>
+        throw InvalidPlanInput()
+    }
+  }
+
+  /**
+   * Parse prototype message of argument list, returns a tuple of (values, types)
+   */
+  def parseArgs(
+      protoArgList: List[proto.RemoteCall.ArgValue],
+      sessionHolder: SessionHolder
+  ): Array[Object] = {
+    val argValues = new Array[Object](protoArgList.size)
+
+    for (i <- 0 until protoArgList.size) {
+      val protoValue = protoArgList(i)
+      argValues(i) = parseArg(protoValue, sessionHolder)
+    }
+    argValues
+  }
+
+  def serializeValue(value: Object, sessionHolder: SessionHolder): proto.RemoteCall.ArgValue = {
+    val protoBuilder = proto.RemoteCall.ArgValue.newBuilder()
+    value match {
+      case v: lang.Integer =>
+        protoBuilder.setInt32Value(v.intValue())
+      case v: lang.Long =>
+        protoBuilder.setInt64Value(v.longValue())
+      case v: lang.Float =>
+        protoBuilder.setFloatValue(v.floatValue())
+      case v: lang.Double =>
+        protoBuilder.setDoubleValue(v.doubleValue())
+      case v: String =>
+        protoBuilder.setStringValue(v)
+      case v: lang.Boolean =>
+        protoBuilder.setBoolValue(v)
+      case v: List[_] =>
+        // TODO
+        throw new UnsupportedOperationException()
+      case v: Map[_, _] =>
+        // TODO
+        throw new UnsupportedOperationException()
+      case v: Object =>
+        val instanceId = sessionHolder.serverSideObjectManager.registerObject(v)
+        protoBuilder.setRemoteObject(
+          RemoteCall.RemoteObject.newBuilder()
+            .setId(instanceId)
+            .build()
+        )
+    }
+    protoBuilder.build()
+  }
+
+  def invokeMethod(
+        instance: Object,
+        methodName: String,
+        argsProto: List[RemoteCall.ArgValue],
+        sessionHolder: SessionHolder): Option[RemoteCall.ArgValue] = {
+    val argValues = parseArgs(argsProto, sessionHolder)
+    val method = instance.getClass.getMethods().find { method =>
```

Comment:
I wonder if this will eventually be a bottleneck for method invocation - maybe not. We could cache results, but that's complicating things

Reply:
For methods that might be called frequently, I can cache return result for them.